namenode/namenode.go

```diff
@@ -44,27 +55,280 @@ func (server *NameNode) Create(path string, success *bool) error {
+			} else {
+				// TODO: mark as stale block and return the info to the datanode
+			}
+		} else {
+			slog.Debug("Create new storage info", "fileName", fileName, "blockID", blockID)
+			inode.storageInfo[blockID] = common.BlockStorageInfo{
+				LatestVersion: version,
+				Size:          metadata.Size,
+				DataNodes:     []string{blockReport.Endpoint},
+			}
+		}
+
+		inode.rwlock.Unlock()
+	}
+
+	// Update the file attributes
+	for fileName, inode := range server.inodes {
+		inode.rwlock.Lock()
+		totalSize := uint(0)
+		for _, storageInfo := range inode.storageInfo {
+			totalSize += storageInfo.Size
+		}
+		inode.fileAttributes.Size = totalSize
+		inode.rwlock.Unlock()
+
+		slog.Debug("Updated inode", slog.String("fileName", fileName),
+			"fileAttributes", inode.fileAttributes, slog.Any("storageInfo", inode.storageInfo))
+	}
+
+	slog.Info("ReportBlock success", "dataNodeEndpoint", blockReport.Endpoint)
+
+	*success = true
+	return nil
+}
+
+func (server *NameNode) Heartbeat(heartbeat common.Heartbeat, success *bool) error {
+	slog.Info("Receive Heartbeat", "dataNodeEndpoint", heartbeat.Endpoint)
+
+	server.inodeRWLock.Lock()
+	defer server.inodeRWLock.Unlock()
+
+	// If the datanode is not registered, ignore the heartbeat
+	if _, exists := server.datanodes[heartbeat.Endpoint]; !exists {
+		*success = false
+		slog.Error("Heartbeat datanode not registered", "dataNodeEndpoint", heartbeat.Endpoint)
+		return errors.New("datanode not registered")
+	}
+
+	// Update the status of the datanode
+	server.datanodes[heartbeat.Endpoint] = DataNodeInfo{
+		heartbeatReceived: true,
+		isAlive:           true,
+	}
+
+	*success = true
+	return nil
+}
+
+func (server *NameNode) heartbeatMonitor() {
+	for {
+		time.Sleep(common.HEARTBEAT_MONITOR_INTERVAL) // Check periodically
+
+		server.datanodeRWLock.Lock()
+		for endpoint, dataNode := range server.datanodes {
+			if !dataNode.heartbeatReceived {
+				// The datanode hasn't sent a heartbeat
+				server.datanodes[endpoint] = DataNodeInfo{
+					heartbeatReceived: false,
+					isAlive:           false,
+				}
+			} else {
+				server.datanodes[endpoint] = DataNodeInfo{
+					heartbeatReceived: false,
+					isAlive:           true,
+				}
+			}
+		}
+		server.datanodeRWLock.Unlock()
+	}
+}
+
+// ReadRequest handles the read request from the client,
+// and returns (Block Name, DataNode Endpoint) pair for each block
+// TODO: only return DataNode Endpoint ?
+func (server *NameNode) ReadRequest(args *common.ReadFileRequest, reply *common.ReadFileResponse) error {
+	slog.Info("ReadRequest", "file", args.FileName)
+
+	server.inodeRWLock.RLock()
+	inode, exists := server.inodes[args.FileName]
+	server.inodeRWLock.RUnlock()
+
+	if !exists {
+		reply.Succeeded = false
+		slog.Error("file not found", "file", args.FileName)
+		return errors.New("file not found")
+	}
+
+	// TODO: if DataNode status changed during request, the client should resend the request
+	dataNodesStatus := server.getDataNodesStatus()
```

Comment:
I can't fully understand why we need this one.

Reply:
I was thinking that to reduce NameNode operations, I designed to fetch the liveness of datanodes only one time outside the loop, so that if a datanode's liveness changed afterwards, we may return a dead datanode to client (but this is inevitable anyway because a datanode can be dead duing network transmission)
This was a note for me to double check the semantic here, I'll remove it in the next commit